### PR TITLE
executor: implement the CLUSTER_LOG retriever

### PR DESCRIPTION
**Pull request by lonng**
### What problem does this PR solve? 

- [x] merge  https://github.com/pingcap/tidb/pull/14018 first

This PR is the subsequent task of https://github.com/pingcap/tidb/pull/14018 and part of issue #13567.

### What is changed and how it works?

Implement the `ClusterLogRetriver`, which is used to search cluster log via SQL.

e.g:
1. Search all `DDL` related log: `select * from cluster_log where message like '%ddl%'` or  `select * from cluster_log where message regexp '.*ddl.*'`
2. Filter log by level: `select * from cluster_log where level='info'`.
3. Search log between time range: `select * from cluster_log where time > '2019-10-14 15:30:39'` and time < '2019-10-19 15:30:39'`.
4. Select log in target cluster component type: `select * from cluster_log where type='tikv'`
5. Select log in target address: `select * from cluster_log where address='192.168.15.2' or  address='192.168.15.2'` or `select * from cluster_log where address in ('192.168.15.2', '192.168.15.2')`.

All the above-mentioned predicates will be pushed down to the memory table to eliminate the log search request amount. e.g: `select * from cluster_log where address in ('192.168.15.2', '192.168.15.2')` only send requests to `'192.168.15.2', '192.168.15.2'` instead of all cluster nodes.

### Check List 

Tests 

 - Unit test
 - Manual test
```
mysql> select * from cluster_log where type='pd' and level='warn' and message like '%';
+-------------------------+------+-----------------+-------+---------------------------------------------------------------------------------------------------------------------------+
| TIME                    | TYPE | ADDRESS         | LEVEL | MESSAGE                                                                                                                   |
+-------------------------+------+-----------------+-------+---------------------------------------------------------------------------------------------------------------------------+
| 2019/12/16 16:28:28.357 | pd   | 127.0.0.1:49901 | Warn  | [store.go:1288] ["simple token is not cryptographically signed"]                                                          |
| 2019/12/16 16:28:29.852 | pd   | 127.0.0.1:49901 | Warn  | [history_buffer.go:138] ["load history index failed"] [error="leveldb: not found"]                                        |
| 2019/12/16 16:29:03.773 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283320320] |
| 2019/12/16 16:29:13.774 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283267072] |
| 2019/12/16 16:29:23.775 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283267072] |
| 2019/12/16 16:29:33.776 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283267072] |
| 2019/12/16 16:29:43.777 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283267072] |
| 2019/12/16 16:29:53.778 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283267072] |
| 2019/12/16 16:30:03.779 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283267072] |
| 2019/12/16 16:30:13.780 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283267072] |
| 2019/12/16 16:30:23.781 | pd   | 127.0.0.1:49901 | Warn  | [cluster.go:383] ["store does not have enough disk space"] [store-id=1] [capacity=1574498656256] [available=280283262976] |
+-------------------------+------+-----------------+-------+---------------------------------------------------------------------------------------------------------------------------+
11 rows in set (0.00 sec)
```

```
mysql> select * from cluster_log where type='pd' and message like '%scheduler%' and message regexp '.*coordinator.*';
+-------------------------+------+-----------------+-------+-----------------------------------------------------------------------------------------+
| TIME                    | TYPE | ADDRESS         | LEVEL | MESSAGE                                                                                 |
+-------------------------+------+-----------------+-------+-----------------------------------------------------------------------------------------+
| 2019/12/16 16:28:56.764 | pd   | 127.0.0.1:49901 | Info  | [coordinator.go:172] ["coordinator starts to run schedulers"]                           |
| 2019/12/16 16:28:56.765 | pd   | 127.0.0.1:49901 | Info  | [coordinator.go:236] ["create scheduler"] [scheduler-name=balance-region-scheduler]     |
| 2019/12/16 16:28:56.766 | pd   | 127.0.0.1:49901 | Info  | [coordinator.go:236] ["create scheduler"] [scheduler-name=balance-leader-scheduler]     |
| 2019/12/16 16:28:56.766 | pd   | 127.0.0.1:49901 | Info  | [coordinator.go:236] ["create scheduler"] [scheduler-name=balance-hot-region-scheduler] |
| 2019/12/16 16:28:56.767 | pd   | 127.0.0.1:49901 | Info  | [coordinator.go:236] ["create scheduler"] [scheduler-name=label-scheduler]              |
+-------------------------+------+-----------------+-------+-----------------------------------------------------------------------------------------+
5 rows in set (0.01 sec)
```

Release note

 - Support search log via SQL, e.g: `SELECT * FROM cluster_log WHERE level='debug' AND message REGEXP '.*coprocessor.*'`

### CODE REVIEW

File: executor/cluster_reader.go [hunk 1]
```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
...
+		// initialize the heap
+		e.heap = &logResponseHeap{}
+		n := len(e.results)
+		for i := n - 1; i >= 0; i-- {
+			result := <-e.results[i]
+			if result.err != nil || len(result.messages) == 0 {
+				if result.err != nil {
+					ctx.GetSessionVars().StmtCtx.AppendWarning(result.err)
+				}
+				e.results = append(e.results[:i], e.results[i+1:]...)
```
Comment (djshow832) [inline executor/cluster_reader.go]: Is `e.results` needed any longer? If not, why modify it?

File: executor/cluster_reader.go [hunk 2]
```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
...
+				continue
+			}
+			*e.heap = append(*e.heap, result)
+		}
+		heap.Init(e.heap)
+	}
+
+	// Merge the results
+	var finalRows [][]types.Datum
+	for e.heap.Len() > 0 && len(finalRows) < 1024 {
```
Comment (djshow832) [inline executor/cluster_reader.go]: Make `1024` configurable.

File: executor/cluster_reader.go [hunk 3]
```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
...
+		min := heap.Pop(e.heap).(logStreamResult)
+		item := min.messages[0]
+		time := time.Unix(item.Time/1000, (item.Time%1000)*int64(time.Millisecond))
+		finalRows = append(finalRows, types.MakeDatums(
+			time.Format("2006/01/02 15:04:05.000"),
```
Comment (djshow832) [inline executor/cluster_reader.go]: I don't understand...

File: executor/cluster_reader.go [hunk 4]
```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
...
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+func (e *clusterLogRetriever) startRetrieving(ctx sessionctx.Context) error {
+	isFailpointTestMode := false
+	serversInfo, err := infoschema.GetClusterServerInfo(ctx)
```
[the remark this answers is not in the document]
Comment (lonng) [reply executor/cluster_reader.go]: The error has been check at the below (give a chance to Failpoint to erase the error).

File: executor/cluster_reader.go [hunk 5]
```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
...
+		EndTime:   e.extractor.EndTime,
+		Levels:    levels,
+		Patterns:  e.extractor.Patterns,
+	}
+
+	// There is no performance isssue to check this variable everytime because it will
+	// be elimitation in non-failpoint mode.
+	if !isFailpointTestMode {
+		if req.StartTime == 0 {
+			req.StartTime = time.Now().UnixNano()/int64(time.Millisecond) - int64(30*time.Minute/time.Millisecond)
```
[the remark this answers is not in the document]
Comment (lonng) [reply executor/cluster_reader.go]: The `SkipRequest` will be true if this situation occurs (See: https://github.com/pingcap/tidb/blob/7262383c8829837e596ab8ff8f8c4ac5f35f09bd/planner/core/memtable_predicate_extractor.go#L459).

File: executor/cluster_reader.go (hunk 1 above)
Comment (lonng) [reply executor/cluster_reader.go]: nice catch.

File: executor/cluster_reader.go (hunk 3 above)
Comment (lonng) [reply executor/cluster_reader.go]: Make the result time has the same format with original log text but don't contain the time zone information.

File: executor/cluster_reader.go (hunk 2 above)
Comment (lonng) [reply executor/cluster_reader.go]: done

File: executor/cluster_reader.go [hunk 6]
```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
...
+	addr     string
+	typ      string
+	messages []*diagnosticspb.LogMessage
+	err      error
+}
+
+type logResponseHeap []logStreamResult
+
+func (h logResponseHeap) Len() int { return len(h) }
+func (h logResponseHeap) Less(i, j int) bool {
```
[the remark this answers is not in the document]
Comment (lonng) [reply executor/cluster_reader.go]: done

File: executor/cluster_reader.go [hunk 7]
```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
...
+	*h = old[0 : n-1]
+	return x
+}
+
+func (e *clusterLogRetriever) startRetrieving(ctx sessionctx.Context) error {
+	isFailpointTestMode := false
+	serversInfo, err := infoschema.GetClusterServerInfo(ctx)
+	failpoint.Inject("mockClusterLogServerInfo", func(val failpoint.Value) {
+		if s := val.(string); len(s) > 0 {
+			serversInfo = serversInfo[:0]
```
[the remark this answers is not in the document]
Comment (lonng) [reply executor/cluster_reader.go]: done

File: executor/cluster_reader.go (hunk 3 above)
Comment (djshow832) [reply executor/cluster_reader.go]: I suggest rename `time`. It overwrites package name `time`.

Comment (lonng) [reply executor/cluster_reader.go]: done

File: executor/cluster_reader.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (djshow832) [reply executor/cluster_reader.go]: No, I mean, you changed `startTime` here, so:
 - If user wants [0, now - 1h], it should return something. 
 - But you changed it to [now - 0.5h, now - 1h], and it returns nothing.

[the remark this answers is not in the document]
Comment (lonng) [reply executor/cluster_reader.go]: nice catch

[the remark this answers is not in the document]
Comment (lonng) [reply executor/cluster_reader.go]: done

**Comment by djshow832** [comment]
LGTM

**Pull request merged**